go/protocol/keybase1/extras.go

```diff
@@ -2764,3 +2780,245 @@ func (fct *FolderConflictType) UnmarshalText(text []byte) error {
 	}
 	return nil
 }
+
+func (h *HiddenTeamChain) Tail() *HiddenTeamChainLink {
+	last := h.Last
+	if last == Seqno(0) {
+		return nil
+	}
+	ret, ok := h.Inner[last]
+	if !ok {
+		return nil
+	}
+	return &ret
+}
+
+func (s Signer) UserVersion() UserVersion {
+	return UserVersion{
+		Uid:         s.U,
+		EldestSeqno: s.E,
+	}
+}
+
+func (p PerTeamSeedCheck) Hash() (*PerTeamSeedCheckPostImage, error) {
+	if p.Version != PerTeamSeedCheckVersion_V1 {
+		return nil, errors.New("can only handle PerTeamKeySeedCheck V1")
+	}
+	ret := sha256.Sum256(p.Value[:])
+	return &PerTeamSeedCheckPostImage{
+		Version: PerTeamSeedCheckVersion_V1,
+		Value:   PerTeamSeedCheckValuePostImage(ret[:]),
+	}, nil
+}
+
+func (p PerTeamSeedCheckPostImage) Eq(p2 PerTeamSeedCheckPostImage) bool {
+	return (p.Version == p2.Version) && hmac.Equal(p.Value[:], p2.Value[:])
+}
+
+func (r HiddenTeamChainRatchet) Flat() []LinkTripleAndTime {
+	var ret []LinkTripleAndTime
+	add := func(p *LinkTripleAndTime) {
+		if p != nil {
+			ret = append(ret, *p)
+		}
+	}
+	add(r.Main)
+	add(r.Blinded)
+	add(r.Self)
+	return ret
+}
+
+func (r HiddenTeamChainRatchet) Max() Seqno {
+	var ret Seqno
+	visit := func(p *LinkTripleAndTime) {
+		if p != nil && p.Triple.Seqno > ret {
+			ret = p.Triple.Seqno
+		}
+	}
+	visit(r.Main)
+	visit(r.Blinded)
+	visit(r.Self)
+	return ret
+}
+
+func (r HiddenTeamChainRatchet) MaxTriple() (ret *LinkTriple) {
+	visit := func(p *LinkTripleAndTime) {
+		if p != nil && (ret == nil || p.Triple.Seqno > ret.Seqno) {
+			ret = &p.Triple
+		}
+	}
+	visit(r.Main)
+	visit(r.Blinded)
+	visit(r.Self)
+	return ret
+}
+
+func (r *HiddenTeamChainRatchet) Merge(r2 HiddenTeamChainRatchet) (updated bool) {
+	visit := func(l1 **LinkTripleAndTime, l2 *LinkTripleAndTime) {
+		if l2 != nil && (*l1 == nil || (*l1).Triple.Seqno < l2.Triple.Seqno) {
+			*l1 = l2
+			updated = true
+		}
+	}
+	visit(&r.Main, r2.Main)
+	visit(&r.Blinded, r2.Blinded)
+	visit(&r.Self, r2.Self)
+	return updated
+}
+
+func (r LinkTripleAndTime) Clashes(r2 LinkTripleAndTime) bool {
+	l1 := r.Triple
+	l2 := r2.Triple
+	return (l1.Seqno == l2.Seqno && l1.SeqType == l2.SeqType && !l1.LinkID.Eq(l2.LinkID))
+}
+
+func (d *HiddenTeamChain) Merge(newData HiddenTeamChain) (updated bool, err error) {
```

Comment:
Might want to make a note in avdl file to consider updating this if the protocol changes